GoogleDataTransportCCTSupport/GDTCCTLibrary/GDTFLLPrioritizer.m

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "GDTCCTLibrary/Private/GDTFLLPrioritizer.h"
+
+#import <GoogleDataTransport/GDTCOREvent.h>
+#import <GoogleDataTransport/GDTCORRegistrar.h>
+#import <GoogleDataTransport/GDTCORStoredEvent.h>
+#import <GoogleDataTransport/GDTCORTargets.h>
+
+const static int64_t kMillisPerDay = 8.64e+7;
+
+@implementation GDTFLLPrioritizer
+
++ (void)load {
+  GDTFLLPrioritizer *prioritizer = [GDTFLLPrioritizer sharedInstance];
+  [[GDTCORRegistrar sharedInstance] registerPrioritizer:prioritizer target:kGDTCORTargetFLL];
+}
+
++ (instancetype)sharedInstance {
+  static GDTFLLPrioritizer *sharedInstance;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    sharedInstance = [[GDTFLLPrioritizer alloc] init];
+  });
+  return sharedInstance;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _queue = dispatch_queue_create("com.google.GDTFLLPrioritizer", DISPATCH_QUEUE_SERIAL);
+    _events = [[NSMutableSet alloc] init];
+  }
+  return self;
+}
+
+#pragma mark - GDTCORPrioritizer Protocol
+
+- (void)prioritizeEvent:(GDTCORStoredEvent *)event {
+  dispatch_async(_queue, ^{
+    [self.events addObject:event];
+  });
+}
+
+- (GDTCORUploadPackage *)uploadPackageWithConditions:(GDTCORUploadConditions)conditions {
+  GDTCORUploadPackage *package = [[GDTCORUploadPackage alloc] initWithTarget:kGDTCORTargetFLL];
+  dispatch_sync(_queue, ^{
+    NSSet<GDTCORStoredEvent *> *logEventsThatWillBeSent;
+    // A high priority event effectively flushes all events to be sent.
+    if ((conditions & GDTCORUploadConditionHighPriority) == GDTCORUploadConditionHighPriority) {
+      package.events = self.events;
+      return;
+    }
+
+    // If on wifi, upload logs that are ok to send on wifi.
+    if ((conditions & GDTCORUploadConditionWifiData) == GDTCORUploadConditionWifiData) {
```

Comment:
Why bitwise-and tests, but no bitwise-or assigns?

Reply:
The upload conditions are computed and assigned in GDT. This is just the same thing that GDTCCT's prioritizer does.

Reply:
I don't see bitwise or's in GDT either. Am I missing where the conditions are merged together?

Reply:
No you're right. Right now, upload conditions are mutually exclusive--I just don't think that'll always be the case. There used to be a condition called "High Priority" that was ORed in when a high priority log was waiting to be sent, and that may be the case again soon with the sqlite refactor.

https://github.com/firebase/firebase-ios-sdk/blob/6be7fc0d7dd047dc26dce1937e55bac2d9050e4d/GoogleDataTransport/GDTCORLibrary/GDTCORUploadCoordinator.m#L134

Reply:
OK thanks. SGTM